### ISSUE TITLE
mqtt client doesn't reconnect when connection was dropped

### ISSUE THREAD

**Issue by nickbr5** [php-mqtt/client#83]
Hi, 

I use this library with a symfony command running with supervisor.

Everything works until I shutdown the mqtt broker for test purpose. If I enable the mqtt broker again, 
supervisor restarts the script and the code triggers the `MqttClient::connect `function again, but for some reason the code hangs in the function `performConnectionHandshake`. In this function there is a while loop where the code hangs.

A difference I see between the initial call and the call after the interruption is that debugging this code in `performConnectionHandshake` returns 1 on the initial call and 0 after the interruption.

```
            // data from the socket.
            $buffer        = $this->readFromSocket(1);

            print strlen($buffer);
```

Any ideas or an example of how to deal with connection lost during a command in supervisor would be appreciated.

Thanks!

**Comment by Namoshek** [comment]
I think your current way of handling a disconnect from the broker (or a connection loss, to be precise) is fine and not the issue. My best guess is that it is related to your connection settings or a potential proxy between client and broker. Can you please provide some details about your setup (broker, client connection settings, TLS settings of the broker, proxy, etc.)?

Just as a side note: the client currently does not handle reconnects on its own. It is an open issue (see #38); not because of the TCP connection but because of the protocol logic which is hard to deal with in this case. Therefore, a clean restart of the script seems reasonable.

**Comment by nickbr5** [comment]
Hi @Namoshek ,

I use rabbitmq running in a docker container as an mqtt broker.
My client connection settings are very basic with only a username & password. No TLS used.
I test the connection loss by stopping the container and start it again.
This is the code I use for connecting:

```
try {
            // Create a new instance of an MQTT client and configure it to use the shared broker host and port.
            $this->client = new MqttClient($_ENV['MQTT_BROKER_HOST'], $_ENV['MQTT_BROKER_PORT'], null, MqttClient::MQTT_3_1, null, $this->logger);

            // Create and configure the connection settings as required.
            $connectionSettings = (new ConnectionSettings)
                ->setUsername($_ENV['MQTT_BROKER_USERNAME'])
                ->setPassword($_ENV['MQTT_BROKER_PASSWORD']);

            // Connect to the broker with the configured connection settings and with a clean session.
            $this->client->connect($connectionSettings, true);
        } catch (MqttClientException $e) {
            echo 'error:'.$e->getMessage();
            // MqttClientException is the base exception of all exceptions in the library. Catching it will catch all MQTT related exceptions.
            $this->logger->critical('Could not connect to MQTT broker: ' . $e->getMessage());
            throw new MqttException('Could not connect to MQTT broker: ' . $e->getMessage());
        }

```

**Comment by Namoshek** [comment]
Are you confident the broker has finished booting when the client attempts to reconnect?

**Comment by nickbr5** [comment]
@Namoshek you were on the right track.
There is no exception thrown by the `MqttClient::connect` function at a certain moment even if the rabbitmq isn't fully booted yet.

**Comment by Namoshek** [comment]
And your client hangs on line 326?

https://github.com/php-mqtt/client/blob/e3cbfcb09868d0d2ce73b2953b91c47fa0ac6876/src/MqttClient.php#L324-L352

If so, then I think this is gonna be hard to solve. Because if the broker accepts the connection (socket is opened) but does not answer requests as expected, the client has no other option than to wait for a response. We can maybe implement a timeout, but not very easily. PHP doesn't really allow setting a timeout on `fread()` calls.

**Comment by nickbr5** [comment]
Once it's connected before the rabbitmq is fully booted it doesn't work anymore.
I tried the code below but still the code hangs.
When I place the `sleep` above establishSocketConnection everything works as expected.

```
try {
            $this->establishSocketConnection();
            sleep(30);
            $this->performConnectionHandshake($useCleanSession);
```
I think the only solution is to know if rabbmitmq is fully loaded, but not sure how.

**Comment by nickbr5** [comment]
@Namoshek I fixed it by waiting for the first byte to be read from the socket:

```
$data = $this->messageProcessor->buildConnectMessage($this->settings, $useCleanSession);

            $this->logger->debug('Sending connection handshake to broker.');

            $this->writeToSocket($data);

            // Start by waiting for the first byte, then using polling logic to fetch all remaining
            // data from the socket.
            $buffer        = $this->readFromSocket(1);

            if(strlen($buffer) === 0) {
                throw new DataTransferException(
                    DataTransferException::EXCEPTION_RX_DATA,
                    'At least 1 byte must be present in the socket.'
                );
            }
```

is this a valid fix? I guess a socket stream will always contain at least 1 byte?

**Comment by Namoshek** [comment]
Does that work for you? If yes, it can only be because the socket is closed by the server. That is because line 326 mentioned above is a blocking operation which will return only after retrieving the first byte. At least to my understanding.

What might work for you is deploying HAProxy in front of the MQTT broker. It can not only act as additional protective layer, it is also able to keep requests from the broker when they are malformed and it supports TCP health checks like connecting to the broker and performing a handshake, to see if it is alive. I don't really have a finished configuration for this at hand right now though.

**Comment by Namoshek** [comment]
I can confirm the issue now, although I'm quite sure it has to do with RabbitMQ being run via Docker. If so, it is not a RabbitMQ issue but rather an issue with Docker binding the socket immediately, even if the application inside the container is not ready to accept connections yet. The message sent as part of the connection handshake is simply dropped. It is a known issue and also the reason why checking for an open port is unrealiable in containerized environments.

That said, I think the best solution to this problem is detecting this type of deadlock and restarting the script again via supervisor. I've created a PR in #84 which potentially addresses the issue (by aborting the connection attempt after the configured connection timeout). Can you please try again using `dev-fix-connect-handshake-timeout` as package version in your `composer.json`?

**Comment by nickbr5** [comment]
Hi @Namoshek,  Just tried branch `dev-fix-connect-handshake-timeout` and I didn't have any problems anymore.
Thanks for investigating the problem and creating this fix!

**Comment by Namoshek** [comment]
Awesome! I've released the fix as (`v1.1.1`](https://github.com/php-mqtt/client/releases/tag/v1.1.1).
